jep/0000/README.adoc

```diff
@@ -0,0 +1,227 @@
+// Uncomment when this JEP status is set to Accepted, Rejected or Withdrawn.
+//| Resolution
+//| :bulb: Link to relevant post in the jenkinsci-dev@ mailing list archives :bulb:
+
+|===
+
+== Abstract
+
+One of the four pillars of Jenkins Essentials is that it should be link:https://github.com/jenkinsci/jep/tree/master/jep/300#sane-defaults[autoconfigured with sane defaults for various cloud platforms]. This specification explains what setup is done to autoconfigure Jenkins Essentials to run in the link:https://aws.amazon.com[Amazon Web Service] environment.
+
+== Specification
+
+=== Global overview
+
+The setup uses an link:https://aws.amazon.com/cloudformation[AWS _CloudFormation_ template] that instantiates the right resources.
+
+It creates an EC2 instance, to host the Jenkins Essentials master in a dedicated Instance Profile.
+The master will be able to provision agents from EC2, and store artifacts directly in S3.
+
+=== Used plugins and configuration
+
+* link:https://plugins.jenkins.io/ec2[EC2 plugin]: to provide dynamically provisioned agents
+  footnote:[even more important for Jenkins Essentials which is configured with the link:JENKINS-49861[sane 0-executor for the master node]].
+* link:https://plugins.jenkins.io/artifact-manager-s3[Artifact Manager on S3 plugin]: to store artifacts and Pipeline stashes in Amazon S3.
+
+// JENKINS-52342 for ^ which is the design, but not done fully yet.
+
+As a general rule for Essentials, we are using the `configuration-as-code` plugin to do the necessary configuration.
+
+*No credentials* are specified.
+The permissions are retrieved dynamically through the instance profile used when creating the EC2 instance for the Essentials master.
+
+=== Instance Profile and IAM role
+
+A dedicated instance profile is created by the _CloudFormation_ template to allow creating EC2 instances for agents, and access a specific S3 bucket to store artifacts and Pipeline stashes.
+
+==== EC2 policy
+
+[source,json]
+"Statement": [{
+    "Sid": "ControlEC2JenkinsAgents",
+    "Effect": "Allow",
+    "Action": [
+        "ec2:DescribeInstances",
+        "ec2:TerminateInstances",
+        "ec2:RequestSpotInstances",
+        "ec2:DeleteTags",
+        "ec2:CreateTags",
+        "ec2:DescribeRegions",
+        "ec2:RunInstances",
+        "ec2:DescribeSpotInstanceRequests",
+        "ec2:StopInstances",
+        "ec2:DescribeSecurityGroups",
+        "ec2:GetConsoleOutput",
+        "ec2:DescribeImages",
+        "ec2:CancelSpotInstanceRequests",
+        "ec2:StartInstances",
+        "ec2:DescribeAvailabilityZones",
+        "ec2:DescribeSubnets",
+        "ec2:DescribeKeyPairs"
+    ],
+    "Resource": "*"
+}]
+
+==== S3 bucket policy
+
+FIXME: refine through JENKINS-52342.
+
+[source,json]
+"Statement": [{
+        "Sid": "TodoRefineSecurityALot",
+        "Effect": "Allow",
+        "Action": [
+            "s3:PutObject",
+            "s3:GetObject",
+            "s3:ListBucket",
+            "s3:DeleteObject"
```

Comment:
You do not need this permission by default: unless overridden with a system property (which I suppose Essentials does not support?), the plugin will not delete artifacts or stashes from S3—that is something that (if even desired) should be configured in S3 itself.

Reply:
Everybody should care about deletions! There's cost, which I agree is cheap, but it's also a liability (IMHO) to unnecessarily maintain older artifacts, especially when developers may be inadvertently archiving data they shouldn't be (nobody ever archives secrets! :smiling_imp: )

I don't see a problem retaining the permission, regardless of the current state of the support by the s3 Artifact Manager plugin.

Reply:
So I'm finally leaving it as-is for a similar reason as the port `50000` allowed above: though the port is not yet used/enabled on the Jenkins Essentials master side itself, we still put that setting as such so that users can enable or disable it easily just by changing config in Jenkins itself.

Essentials is critically about simplicity, so here it's in my opinion the winning criterion to keep such settings this way, provided obviously we agree it's not a glaring wide security hole (as I think it is not, here).

Going to add a blurb about it in the spec to keep track of this, not just as a comment here.